### DISCUSSION TITLE
Welcome!

### DISCUSSION THREAD

**Discussion by fwillett** [fwillett/speechBCI#2] [Announcements]
Hello all! This forum is for discussing the [speech decoding competition](https://eval.ai/web/challenges/challenge-page/2099/overview), [paper](https://www.biorxiv.org/content/10.1101/2023.01.21.524489v2.abstract) or anything in this code repository. Feel free to ask questions and we'll do our best to answer!

**Comment by Farzaneh-Taleb** [comment]
Thanks for sharing the resources. That would be amazing if you could provide a working env variable too!

**Reply by cffan** [reply]
We'd like to help, but could you clarify which `working env variable` is missing?

**Comment by Farzaneh-Taleb** [comment]
I think line 9 in makeTFRecordsFromSession.py should be changed to "from **NeuralDecoder**.neuralDecoder.datasets.speechDataset import PHONE_DEF, VOWEL_DEF, CONSONANT_DEF, SIL_DEF, PHONE_DEF_SIL
"

**Reply by cffan** [reply]
Thanks for the suggestion. Both `from NeuralDecoder.neuralDecoder` and `from neuralDecoder` are valid from our test. If one is preferred over the other, please let us know the reason and we can change it.

**Comment by danielwang365** [comment]
Hi,

I was wondering if was possible to use some other language model decoder for the challenge. The notebook required to build the language model and WFST decoding graph says one needs at least 230 GB of free disk space and 100 GM of RAM. I don't have access to these resources, so I was wondering how I would go about building the language model.

Thanks!

**Reply by fwillett** [reply]
Hi Daniel, thanks for your question! You don't have to build the language model yourself, we released the language model files for the 3-gram and the 5-gram model [here ](https://datadryad.org/stash/dataset/doi:10.5061/dryad.x69p8czpq) - see the languageModel.tar.gz and the languageModel_5gram.tar.gz files. It should be possible to run inference with the 3-gram model with ~60 GB of RAM.

**Comment by ebrahimfeghhi** [comment]
Hi, thank you so much for sharing this dataset! I had a quick question regarding the delay period. In Table 2 in the supplementary section, the delay times for sentences are listed as being between 4 and 5 seconds. However, when I look at the delay times for real time decoding sessions using "trialDelayTimes", I find they range between 3 and 4 seconds. Also, there are times when the delay time computed using "trialState" doesn't align with "trialDelayTimes". For instance, in session "07.27" in the sentences data, the delay time listed at index 321 (Python) for "trialDelayTimes" is 3.04 seconds, whereas the delay time I compute for index 321 using "trialState" is 1.92 seconds (1.92 seconds also aligns with the start and end times from "delayTrialEpochs"). Any clarification would be appreciated, thanks in advance!

**Reply by ebrahimfeghhi** [reply]
Also, trialState sometimes takes on the value "3". Does this refer to something, or should we ignore neural data which aligns to a trialState of "3"? Thanks!

**Reply by fwillett** [reply]
Hi Ebrahim! Thanks for these questions. You're right about Table 2 being wrong - agreed that the delay times range between 3 and 4 seconds, not 4 and 5. Apologies for the mistake. 

Note that trialState takes on a value of "3" during closed-loop trials, and it represents and end-of-trial state where the software plays audio of the decoded sentence using Google Text-to-Speech. The trial remains in this state until the participant presses a button to advance to the next trial.

It looks like the mismatched delay times between the "trialDelayTimes" variable and the "trialState" variable is more complex, and is related to how we aligned the neural data to the task data. For reasons I don't fully understand, sometimes during trialState 3, the task "stutters" and the frame rate drops. It may have something to do with the Text-to-Speech loading. When this happens, it creates imprecision in how task data is aligned to the neural data, and so the time when trialState 3 ends and the delay period of the next trial starts becomes incorrect. Unfortunately it may be complex and not fully possible to recover the exact timing. If you need precise timing for when the delay period begins, I recommend not using trials where the "trialDelayTimes" duration doesn't match the "trialSate" delay period duration, which should still give you a lot of trials. If you do use all the trials, the imprecision should be on the order of about half of the difference in "trialDelayTimes" and "trialState" for that trial. 

My apologies for this bug, and thank you for bringing this to everyone's attention!

**Reply by ebrahimfeghhi** [reply]
Hi Frank! Thank you so much for the response and the clarifications. Okay that makes sense, I'll follow your suggestion and ignore trials where the trialDelayTimes duration doesn't align with the trialState duration. I appreciate it!
